In [5]:
import pandas as pd
import geopandas as gpd
import os
os.getcwd()


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "C:\Users\Jerm\anaconda3\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\Jerm\anaconda3\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\Jerm\anaconda3\lib\site-packages\ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "C:\Users\Jerm\anaconda3\lib\site-packages\traitlets\config\application.py", line 846, in launch_instance
    app.start()
  File "C:\Users\Jerm\anaconda3\lib\site-packages\

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "C:\Users\Jerm\anaconda3\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\Jerm\anaconda3\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\Jerm\anaconda3\lib\site-packages\ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "C:\Users\Jerm\anaconda3\lib\site-packages\traitlets\config\application.py", line 846, in launch_instance
    app.start()
  File "C:\Users\Jerm\anaconda3\lib\site-packages\

AttributeError: _ARRAY_API not found

'C:\\Users\\Jerm\\OneDrive\\AI project\\Dementia\\singapore-dementia-analysis\\notebooks'

In [6]:
demand_df = pd.read_csv('../data/processed/demand_surface.csv')
facilities_df = pd.read_csv('../data/processed/facilities_with_planning_area.csv')
gdf = gpd.read_file('../data/raw/DataMasterPlan2019PlngAreaBoundaryNoSea.geojson')


In [7]:
# Inspect demand surface
print("DEMAND SURFACE")
print(demand_df.shape)
print(demand_df.dtypes)
demand_df.head()

DEMAND SURFACE
(55, 2)
PLN_AREA_N      object
pwd_estimate     int64
dtype: object


,PLN_AREA_N,pwd_estimate
0,ANG MO KIO,5507
1,BEDOK,8197
2,BISHAN,2530
3,BOON LAY,0
4,BUKIT BATOK,3316


In [8]:
# Inspect facilities
print("FACILITIES")
print(facilities_df.shape)
print(facilities_df.dtypes)
facilities_df.head()

FACILITIES
(148, 7)
facility_id        int64
facility_name     object
full_address      object
postal_code        int64
latitude         float64
longitude        float64
PLN_AREA_N        object
dtype: object


,facility_id,facility_name,full_address,postal_code,latitude,longitude,PLN_AREA_N
0,1,4S Eldercare Centre @ Eunos Crescent,31A Eunos Crescent #06-01 Singapore 401031,401031,1.320034,103.901616,GEYLANG
1,2,AWWA Dementia Day Care Centre,123 Ang Mo Kio Avenue 6 #01-4035 Singapore 560123,560123,1.370481,103.844806,ANG MO KIO
2,3,AWWA Dementia Day Care Centre @ Yishun,740 Yishun Avenue 5 #01-490 Singapore 760740,760740,1.430124,103.832153,YISHUN
3,4,Active Global @ Fernvale Glades,460 Sengkang West Way #01-02 Singapore 790460,790460,1.394261,103.869619,SENGKANG
4,5,Active Global Ghim Moh Active Ageing Centre (C...,31A Ghim Moh Link #01-11 Singapore 272031,272031,1.309329,103.784354,QUEENSTOWN


In [9]:
# Inspect GeoJSON
print("GEOJSON")
print(gdf.shape)
print(gdf.dtypes)
gdf.head()

GEOJSON
(55, 11)
OBJECTID         int32
PLN_AREA_N      object
PLN_AREA_C      object
CA_IND          object
REGION_N        object
REGION_C        object
INC_CRC         object
FMEL_UPD_D      object
SHAPE.AREA     float64
SHAPE.LEN      float64
geometry      geometry
dtype: object


,OBJECTID,PLN_AREA_N,PLN_AREA_C,CA_IND,REGION_N,REGION_C,INC_CRC,FMEL_UPD_D,SHAPE.AREA,SHAPE.LEN,geometry
0,1,BEDOK,BD,N,EAST REGION,ER,5F00E6FF084F3364,20191223152014,2.173397e+07,21864.234336,"POLYGON ((103.93208 1.30555, 103.93208 1.30555..."
1,2,BOON LAY,BL,N,WEST REGION,WR,C96AED188C00B2FC,20191223152014,8.282808e+06,18542.688792,"POLYGON ((103.72042 1.32824, 103.72066 1.32865..."
2,3,BUKIT BATOK,BK,N,WEST REGION,WR,3BEC4C829160F28A,20191223152014,1.114018e+07,15236.442232,"POLYGON ((103.76408 1.37001, 103.76385 1.37036..."
3,4,BUKIT MERAH,BM,N,CENTRAL REGION,CR,4850795BB0B6A4F7,20191223152014,1.446120e+07,29161.409834,"POLYGON ((103.82361 1.26018, 103.82368 1.26018..."
4,5,BUKIT PANJANG,BP,N,WEST REGION,WR,656F87D23D6DAB02,20191223152014,9.019940e+06,15891.853279,"POLYGON ((103.77445 1.39029, 103.77444 1.39029..."


In [10]:
print(demand_df['PLN_AREA_N'].sort_values().unique())
print(facilities_df['PLN_AREA_N'].sort_values().unique())
print(gdf['PLN_AREA_N'].sort_values().unique())

['ANG MO KIO' 'BEDOK' 'BISHAN' 'BOON LAY' 'BUKIT BATOK' 'BUKIT MERAH'
 'BUKIT PANJANG' 'BUKIT TIMAH' 'CENTRAL WATER CATCHMENT' 'CHANGI'
 'CHANGI BAY' 'CHOA CHU KANG' 'CLEMENTI' 'DOWNTOWN CORE' 'GEYLANG'
 'HOUGANG' 'JURONG EAST' 'JURONG WEST' 'KALLANG' 'LIM CHU KANG' 'MANDAI'
 'MARINA EAST' 'MARINA SOUTH' 'MARINE PARADE' 'MUSEUM' 'NEWTON'
 'NORTH-EASTERN ISLANDS' 'NOVENA' 'ORCHARD' 'OUTRAM' 'PASIR RIS'
 'PAYA LEBAR' 'PIONEER' 'PUNGGOL' 'QUEENSTOWN' 'RIVER VALLEY' 'ROCHOR'
 'SELETAR' 'SEMBAWANG' 'SENGKANG' 'SERANGOON' 'SIMPANG' 'SINGAPORE RIVER'
 'SOUTHERN ISLANDS' 'STRAITS VIEW' 'SUNGEI KADUT' 'TAMPINES' 'TANGLIN'
 'TENGAH' 'TOA PAYOH' 'TUAS' 'WESTERN ISLANDS' 'WESTERN WATER CATCHMENT'
 'WOODLANDS' 'YISHUN']
['ANG MO KIO' 'BEDOK' 'BISHAN' 'BUKIT BATOK' 'BUKIT MERAH' 'BUKIT PANJANG'
 'CHOA CHU KANG' 'CLEMENTI' 'DOWNTOWN CORE' 'GEYLANG' 'HOUGANG'
 'JURONG EAST' 'JURONG WEST' 'KALLANG' 'MARINE PARADE' 'NOVENA' 'OUTRAM'
 'PASIR RIS' 'PUNGGOL' 'QUEENSTOWN' 'SEMBAWANG' 'SENGKANG' 'SERANGOON'


In [11]:
demand_set = set(demand_df['PLN_AREA_N'])
facilities_set = set(facilities_df['PLN_AREA_N'])
geo_set = set(gdf['PLN_AREA_N'])

print("In facilities but not in demand:", facilities_set - demand_set)
print("In GeoJSON but not in demand:", geo_set - demand_set)
print("In demand but not in GeoJSON:", demand_set - geo_set)

In facilities but not in demand: set()
In GeoJSON but not in demand: set()
In demand but not in GeoJSON: set()


In [12]:
facility_counts = facilities_df.groupby('PLN_AREA_N')['facility_id'].agg('count').reset_index()
print(facility_counts.shape)
facility_counts.head()

(27, 2)


,PLN_AREA_N,facility_id
0,ANG MO KIO,7
1,BEDOK,12
2,BISHAN,3
3,BUKIT BATOK,8
4,BUKIT MERAH,10


In [13]:
merged_df = demand_df.merge(facility_counts, on='PLN_AREA_N', how='left')
merged_df = merged_df.rename(columns={'facility_id': 'facility_count'})
merged_df['facility_count'] = merged_df['facility_count'].fillna(0).astype(int)

print(merged_df.shape)
merged_df.head(10)

(55, 3)


,PLN_AREA_N,pwd_estimate,facility_count
0,ANG MO KIO,5507,7
1,BEDOK,8197,12
2,BISHAN,2530,3
3,BOON LAY,0,0
4,BUKIT BATOK,3316,8
5,BUKIT MERAH,5313,10
6,BUKIT PANJANG,2715,3
7,BUKIT TIMAH,2215,0
8,CENTRAL WATER CATCHMENT,0,0
9,CHANGI,44,0


## Design Note: Aggregated vs Facility-Level Master Table

We are deliberately working at planning area level for this notebook.
The 148-row facility-level source (`facilities_with_planning_area.csv`) 
is preserved as the row-level source of truth for any future drill-down, 
ABM inputs, or stress testing.

A facility-level master table (one row per facility, all planning area 
context columns repeated) is a valid alternative architecture and would 
simplify certain future queries. That refactor is deferred — the current 
pipeline serves the choropleth dashboard output cleanly.

In [14]:
merged_df = merged_df.merge(gdf[['PLN_AREA_N', 'geometry']], on='PLN_AREA_N', how='left')

merged_df.head()

,PLN_AREA_N,pwd_estimate,facility_count,geometry
0,ANG MO KIO,5507,7,"POLYGON ((103.85721 1.39654, 103.85701 1.39672..."
1,BEDOK,8197,12,"POLYGON ((103.93208 1.30555, 103.93208 1.30555..."
2,BISHAN,2530,3,"POLYGON ((103.84297 1.36429, 103.84228 1.36435..."
3,BOON LAY,0,0,"POLYGON ((103.72042 1.32824, 103.72066 1.32865..."
4,BUKIT BATOK,3316,8,"POLYGON ((103.76408 1.37001, 103.76385 1.37036..."


In [15]:
merged_df['pwd_per_facility'] = (merged_df['pwd_estimate'] / merged_df['facility_count']).where(merged_df['facility_count'] > 0)

merged_df

,PLN_AREA_N,pwd_estimate,facility_count,geometry,pwd_per_facility
0,ANG MO KIO,5507,7,"POLYGON ((103.85721 1.39654, 103.85701 1.39672...",786.714286
1,BEDOK,8197,12,"POLYGON ((103.93208 1.30555, 103.93208 1.30555...",683.083333
2,BISHAN,2530,3,"POLYGON ((103.84297 1.36429, 103.84228 1.36435...",843.333333
3,BOON LAY,0,0,"POLYGON ((103.72042 1.32824, 103.72066 1.32865...",NaN
4,BUKIT BATOK,3316,8,"POLYGON ((103.76408 1.37001, 103.76385 1.37036...",414.500000
5,BUKIT MERAH,5313,10,"POLYGON ((103.82361 1.26018, 103.82368 1.26018...",531.300000
6,BUKIT PANJANG,2715,3,"POLYGON ((103.77445 1.39029, 103.77444 1.39029...",905.000000
7,BUKIT TIMAH,2215,0,"POLYGON ((103.79766 1.34813, 103.79708 1.34854...",NaN
8,CENTRAL WATER CATCHMENT,0,0,"POLYGON ((103.80702 1.41126, 103.80662 1.41239...",NaN
9,CHANGI,44,0,"POLYGON ((103.98236 1.39356, 103.98235 1.39357...",NaN


In [16]:
merged_df['no_facility'] = merged_df['facility_count'] == 0
merged_df.head()

,PLN_AREA_N,pwd_estimate,facility_count,geometry,pwd_per_facility,no_facility
0,ANG MO KIO,5507,7,"POLYGON ((103.85721 1.39654, 103.85701 1.39672...",786.714286,False
1,BEDOK,8197,12,"POLYGON ((103.93208 1.30555, 103.93208 1.30555...",683.083333,False
2,BISHAN,2530,3,"POLYGON ((103.84297 1.36429, 103.84228 1.36435...",843.333333,False
3,BOON LAY,0,0,"POLYGON ((103.72042 1.32824, 103.72066 1.32865...",NaN,True
4,BUKIT BATOK,3316,8,"POLYGON ((103.76408 1.37001, 103.76385 1.37036...",414.500000,False


In [17]:
#average pwd per facility 

pwd_per_facility_sort = merged_df[merged_df['facility_count'] > 0][['PLN_AREA_N', 'pwd_estimate', 'facility_count', 'pwd_per_facility']].sort_values('pwd_per_facility', ascending=False)
pwd_per_facility_sort

,PLN_AREA_N,pwd_estimate,facility_count,pwd_per_facility
23,MARINE PARADE,1573,1,1573.000000
16,JURONG EAST,2101,2,1050.500000
11,CHOA CHU KANG,3036,3,1012.000000
6,BUKIT PANJANG,2715,3,905.000000
34,QUEENSTOWN,3591,4,897.750000
40,SERANGOON,3455,4,863.750000
2,BISHAN,2530,3,843.333333
0,ANG MO KIO,5507,7,786.714286
39,SENGKANG,3701,5,740.200000
33,PUNGGOL,2164,3,721.333333


In [18]:
#Population with no facility. undersupply list

dd_no_ss_sort = merged_df[
    (merged_df['no_facility'] == True) & 
    (merged_df['pwd_estimate'] > 0)
][['PLN_AREA_N', 'pwd_estimate', 'facility_count', 'pwd_per_facility']].sort_values('pwd_estimate', ascending=False)

dd_no_ss_sort

,PLN_AREA_N,pwd_estimate,facility_count,pwd_per_facility
7,BUKIT TIMAH,2215,0,NaN
47,TANGLIN,540,0,NaN
36,ROCHOR,465,0,NaN
35,RIVER VALLEY,215,0,NaN
25,NEWTON,205,0,NaN
48,TENGAH,58,0,NaN
20,MANDAI,44,0,NaN
9,CHANGI,44,0,NaN
42,SINGAPORE RIVER,43,0,NaN
45,SUNGEI KADUT,28,0,NaN


In [19]:
#No population, No facilities

No_pop_no_fac = merged_df[
    (merged_df['no_facility'] == True) & 
    (merged_df['pwd_estimate'] == 0)
][['PLN_AREA_N', 'pwd_estimate', 'facility_count', 'pwd_per_facility']]


No_pop_no_fac

,PLN_AREA_N,pwd_estimate,facility_count,pwd_per_facility
3,BOON LAY,0,0,NaN
8,CENTRAL WATER CATCHMENT,0,0,NaN
10,CHANGI BAY,0,0,NaN
19,LIM CHU KANG,0,0,NaN
21,MARINA EAST,0,0,NaN
22,MARINA SOUTH,0,0,NaN
26,NORTH-EASTERN ISLANDS,0,0,NaN
31,PAYA LEBAR,0,0,NaN
32,PIONEER,0,0,NaN
41,SIMPANG,0,0,NaN


In [20]:
#No Population, with facilities

No_pop_withfac = merged_df[
    (merged_df['no_facility'] == False) & 
    (merged_df['pwd_estimate'] == 0)
][['PLN_AREA_N', 'pwd_estimate', 'facility_count', 'pwd_per_facility']]


No_pop_withfac

,PLN_AREA_N,pwd_estimate,facility_count,pwd_per_facility


In [21]:
#pwd largest demand

merged_df[
    (merged_df['no_facility'] == False) & 
    (merged_df['pwd_estimate'] > 0)
][['PLN_AREA_N', 'pwd_estimate', 'facility_count', 'pwd_per_facility']].sort_values('pwd_estimate', ascending=False)

,PLN_AREA_N,pwd_estimate,facility_count,pwd_per_facility
1,BEDOK,8197,12,683.083333
46,TAMPINES,5749,10,574.900000
15,HOUGANG,5509,9,612.111111
0,ANG MO KIO,5507,7,786.714286
5,BUKIT MERAH,5313,10,531.300000
17,JURONG WEST,4691,9,521.222222
49,TOA PAYOH,4484,9,498.222222
54,YISHUN,4313,7,616.142857
53,WOODLANDS,3817,6,636.166667
39,SENGKANG,3701,5,740.200000


In [22]:
# Option 1: mean of ratios
avg_pwd_per_facility = pwd_per_facility_sort['pwd_per_facility'].mean()
print(f"Mean pwd_per_facility across served areas: {avg_pwd_per_facility:.1f}")

Mean pwd_per_facility across served areas: 667.3


In [23]:
total_pwds = pwd_per_facility_sort['pwd_estimate'].sum()
total_facilities = pwd_per_facility_sort['facility_count'].sum()
weighted_avg = total_pwds / total_facilities

print(f"Total PWDs in served areas: {total_pwds}")
print(f"Total facilities: {total_facilities}")
print(f"Weighted avg PWDs per facility: {weighted_avg:.1f}")

Total PWDs in served areas: 91740
Total facilities: 148
Weighted avg PWDs per facility: 619.9


In [24]:
total_unserved_pwds = dd_no_ss_sort['pwd_estimate'].sum()
print(f"Total PWDs in planning areas with no facility: {total_unserved_pwds}")

Total PWDs in planning areas with no facility: 3907


In [25]:
from openpyxl import Workbook

# Drop geometry for Excel - not compatible
merged_excel = merged_df.drop(columns=['geometry'])

# Write to Excel with multiple sheets
with pd.ExcelWriter('../data/processed/dementia_analysis.xlsx', engine='openpyxl') as writer:
    merged_excel.to_excel(writer, sheet_name='Full Analysis', index=False)
    pwd_per_facility_sort.to_excel(writer, sheet_name='Served Areas', index=False)
    dd_no_ss_sort.to_excel(writer, sheet_name='Undersupply', index=False)
    No_pop_no_fac.to_excel(writer, sheet_name='Noise', index=False)

print("Export complete")

Export complete


In [26]:
merged_df.head()

,PLN_AREA_N,pwd_estimate,facility_count,geometry,pwd_per_facility,no_facility
0,ANG MO KIO,5507,7,"POLYGON ((103.85721 1.39654, 103.85701 1.39672...",786.714286,False
1,BEDOK,8197,12,"POLYGON ((103.93208 1.30555, 103.93208 1.30555...",683.083333,False
2,BISHAN,2530,3,"POLYGON ((103.84297 1.36429, 103.84228 1.36435...",843.333333,False
3,BOON LAY,0,0,"POLYGON ((103.72042 1.32824, 103.72066 1.32865...",NaN,True
4,BUKIT BATOK,3316,8,"POLYGON ((103.76408 1.37001, 103.76385 1.37036...",414.500000,False
